### ISSUE TITLE
Permissions/Image issues on The Forge.

### ISSUE THREAD

**Issue by theshipening** [Argonius-Angelus/lancer-clocks#14]
I am subject to this known issue : 
> Due to how the module handles custom themes, users require "Use File Browser" permissions to even open the ClockSheet. This is deemed an edge-case scenario as this module is mainly GM-facing in its functions.
Except that I am the GM and owner of the world, so I'm pretty sure I should have all the permissions

**Screenshots**
![image](https://github.com/Argonius-Angelus/lancer-clocks/assets/140340386/4008b04c-48cb-49bb-9496-4774f0d6c3db)

I'm using the Forge, with Foundry v11 13 315. Uninstalling/installing and deactivating/activating the module. didn't solve the issue

**Comment by abacus-tabletop-games** [comment]
I am the GM in my game and have the same problem. Using Foundry v11.315 and MoSh Unofficial Mothership v0.5.15.

**Comment by Argonius-Angelus** [comment]
Gotta love when Github doesn't tell me when people make issues on my project.
Unfortunately I am incapable of reproducing your specific issue, as I do not use Forge. I wonder if something happened around this time where you are not able to see the default theme locations even as a GM given #15 is reporting a similar issue.

Have you since figured out a workaround? Forge compatibility has always been tenable at best as I do not know what their file structure is like.

**Comment by Argonius-Angelus** [comment]
Looking back at this just now. I am genuinely curious what is going on here. Normally, you don't even get to the stage of being able to open the sheet itself with how this error usually works. What on earth is Forge doing that is causing this?

**Comment by Qenthel94** [comment]
> Looking back at this just now. I am genuinely curious what is going on here. Normally, you don't even get to the stage of being able to open the sheet itself with how this error usually works. What on earth is Forge doing that is causing this?

I had the same issue in forge as a gm. but updating the  "extra lancer clocks" path from "lancer-clocks" to "modules/lancer-clocks" has partially fixed it for me. The image displays, but from time to time it still throws "Failed to load resource: the server responded with a status of 404 ()"

looks like its going by this path when updating the clock: undefined/6clock_4.png 404 (Not Found)

![Image](https://github.com/user-attachments/assets/f41b6d93-6eb5-4f97-bc22-795bf9dd4350)

![Image](https://github.com/user-attachments/assets/a46cdd48-da56-4185-9307-3ba71f9a7a92)

**Comment by Argonius-Angelus** [comment]
That is very weird and very baffling. I wish I could properly test what Forge is doing in order to better understand why it's acting the way it is. Is there some sort of architectural change to how the FilePicker handles things in Forge?

> > Looking back at this just now. I am genuinely curious what is going on here. Normally, you don't even get to the stage of being able to open the sheet itself with how this error usually works. What on earth is Forge doing that is causing this?
> 
> I had the same issue in forge as a gm. but updating the "extra lancer clocks" path from "lancer-clocks" to "modules/lancer-clocks" has partially fixed it for me. The image displays, but from time to time it still throws "Failed to load resource: the server responded with a status of 404 ()"
> 

Which module version and Foundry version are you on for this? There was a recent update that changed things to v12 and I would not be surprised if things broke from the changes made.

**Comment by Argonius-Angelus** [comment]
Ok hold on I just properly noticed what you have listed there for the custom themes, and that shouldn't be discovering any themes. Is something about the fact that "lancer-clocks" exists within the root of the data path breaking the module? The more I learn about how The Forge interacts with my module, the more cursed it sounds.

**Comment by Narshero** [comment]
I think I'm seeing this issue on Forge as well, but it's... even weirder somehow? It seems like there may be some sort of race condition going on; I have a number of custom clock themes I've made myself, and the behavior I'm seeing is that, on any given modification of the clock actor (incrementing/decrementing/resetting the counter, changing the theme or number of segments), one of two things will happen, apparently at random:

- Both the default and custom themes appear in the theme menu, the image path is correctly set to modules/lancer-clocks/themes/[...], both the token and sheet show the right graphic:
![Image](https://github.com/user-attachments/assets/2f785565-4eb5-4d15-89a3-ec2121878ca8)
- Only the custom themes appear in the theme menu, the selected theme lists the first of my custom themes (but doesn't actually change to that theme behind the scenes as far as I can tell), the image path is set to "undefined/[filename]", the token displays the default actor image and the image in the sheet is broken:
![Image](https://github.com/user-attachments/assets/26c4f6bd-5608-4155-9854-c3596ae9f7ce)

Notably, if I use one of the custom themes, I don't see any issues, everything loads and displays fine.

Running Foundry V12 Stable Build 331, Lancer 2.6.0, Lancer Clocks 1.0.29

**Comment by Argonius-Angelus** [comment]
So to confirm, the default themes tend to break, but custom themes work consistently for you? What do you have listed as the custom theme directory and how is that reflected in the file structure on The Forge's side of things?

**Comment by Argonius-Angelus** [comment]
>     * Only the custom themes appear in the theme menu, the selected theme lists the first of my custom themes (but doesn't actually change to that theme behind the scenes as far as I can tell), the image path is set to "undefined/[filename]", the token displays the default actor image and the image in the sheet is broken:

This specific behavior is also something I noticed, and seems to be a behavior specific to the `{{#selectOptions}}` handlebar helper, wherein it doesn't actually update what is being sent until you change the options. As soon as you change it, it should fix itself.

**Comment by Narshero** [comment]
> So to confirm, the default themes tend to break, but custom themes work consistently for you? What do you have listed as the custom theme directory and how is that reflected in the file structure on The Forge's side of things?

Yes, the defaults are breaking intermittently but the custom themes are working fine. I have my custom theme directory set to the default lancer-clocks, which The Forge puts in their "Assets Library" portion of the file system; I have no idea where that exists in relation to the "User Data" or "The Bazaar" portions, possibly they're all just piled on top of each other and their plugin just shows/hides what's in each category.

(Actually, given there appears to be a /modules/lancer-clocks/ in both "User Data" and "The Bazaar", I wonder if there might be some sort of conflict where the module is getting one or the other at random and can load from one but not the other. Might explain why the failure is intermittent for me.)

![Image](https://github.com/user-attachments/assets/32a9001a-d073-4f0e-8eec-fe5122bf6529)

**Comment by Argonius-Angelus** [comment]
This is cursed as all hell. I'll add a note to Known Issues about custom themes fixing things at the very least. I'll also see if I can't get some help from The Forge's staff to understand why things are acting the way they are.

**Comment by Argonius-Angelus** [comment]
Oh boy it's a race condition.
This will take a while to fix as I might need to completely overhaul the way clocks are constructed to play well with Forge.

**Comment by Argonius-Angelus** [comment]
Ok got some good news and bad news.

Good news:
I fixed the race condition and base themes no longer break as a result.

Bad news:
Clocks are now very slow in Forge specifically (and likely will take a performance hit for other externally-hosted games). Next to no performance hit has been observed when using a Clock actor on a self-hosted instance of Foundry.

**Comment by Argonius-Angelus** [comment]
I have pushed out an update that also includes options to offset the slowdown of these changes for Forge users. I will be keeping this issue open however until I have exhausted all other possibilities of improving performance.
